### ISSUE TITLE
Rewrite messaging layer

### ISSUE THREAD

**Issue by volovyks** [sig-net/mpc#52]
### Description
Our messaging layer is one of the bottlenecks and does not follow industry standards.
Sequential (blocking) REST calls are causing low throughput, especially when the system is under load.

The maximum number of messages we can send is only 2-3X from what we are sending when the system is not under load:
<img width="592" alt="image" src="https://github.com/user-attachments/assets/c4665dc8-6c11-4194-895e-3515c37af23f">

Median latency grows significantly, ~5-10X.
<img width="605" alt="image" src="https://github.com/user-attachments/assets/58cb67b8-a160-4866-ac99-fb0c680e34aa">

The most obvious choice is [libp2p](https://libp2p.io/) and its [Rust implementation](https://docs.rs/libp2p/latest/libp2p/). It is used in many decentralized systems, including Ethereum. But we should consider alternatives, such as [QUIC](https://quicwg.org/) ([rust impl](https://github.com/quinn-rs/quinn)).

**Comment by ailisp** [comment]
In early days of near protocol there was a discussion of using libp2p (c++), but the concludion was it's over engineered but not robust enough

**Comment by ppca** [comment]
not suggesting libraries, but a current limitation of our implementation: our number of messages in message queue is capped by `let (sender, receiver) = mpsc::channel(16384);` now. Whenever receiver stops receiving new messages, the channel will be full and no message sending can be successful.
Not sure if other libraries could better manage this.

**Comment by volovyks** [comment]
Updated Grafana metrics show that when the system is under load, each node is sending ~40 messages/second, which is not much. Communication channel capacity depends on many factors, but `libp2p` with TCP or QUIC beneath it can give us up to 100,000 messages/second.
<img width="745" alt="image" src="https://github.com/user-attachments/assets/20d4298b-ebaf-4a88-9b57-2cc27daf0f32">

**Comment by volovyks** [comment]
When choosing a way forward it is important to ensure it is good for future horizontal scaling. We may have many triple (possibly pre-signature and signature) generation nodes managed by each node provider.

**Comment by jakmeier** [comment]
Yeah, I think eventually we want to move to a different messaging layer. However, it might make sense to identify the bottlenecks of the current code more clearly before moving to a different messaging layer.

True, using HTTP is not as efficient as using QUIC directly or building on top of libp2p but we shouldn't expect a 100x improvement in throughput or latency. If that's the order of magnitude what we are aiming for, probably we need to fix how and when we send messages, not the messaging layer itself.

> Our messaging layer is one of the bottlenecks and does not follow industry standards.
> Sequential (blocking) REST calls are causing low throughput, especially when the system is under load.

Hm, I wonder how much the throughput really is limited by the messaging layer. As far as I can tell, the code already uses the async [`reqwest::Client`](https://docs.rs/reqwest/0.11.27/reqwest/struct.Client.html). So, the messaging layer itself should be able to run things concurrently, if used correctly.

@volovyks  When you say we have sequential blocking REST calls, did you have specific requests in mind? If yes, can you please point me to the code locations so I can check if there are any issues in how we use reqwest?

> Updated Grafana metrics show that when the system is under load, each node is sending ~40 messages/second, which is not much. Communication channel capacity depends on many factors, but libp2p with TCP or QUIC beneath it can give us up to 100,000 messages/second.

Pretty sure reqwest is capable of much higher throughput 40 msg / sec. We should probably identify why exactly we are not sending more under load. Are we at 100% CPU utilization? Are we waiting on a lock held by a different thread? Or is the cryptographic algorithm itself not emitting more requests?

**Comment by volovyks** [comment]
**Sequential blocking REST calls:**
I was refering to the fact that we are keeping a deque of messages in  [MessageQueue](https://github.com/sig-net/mpc/blob/4c32d0038ed2cb6699e26cadd2327af3bce5e696/chain-signatures/node/src/http_client.rs#L97) struct. Which is wrapped in `Arc` and `RwLock` and used all ower the project. Both `push()` and `send_encrypted()` functions can block each other. But @ChaoticTempest is applying a new multithreaded design which is changing how we add and send messages.

**Are we at 100% CPU utilization?**
No, the first 25% of the time here is active triple generation, which is barely visible.
<img width="744" alt="image" src="https://github.com/user-attachments/assets/7d62c6df-6a9b-4337-b95d-ae4430e3892c" />

@ChaoticTempest is changing a lot of logic now. We will try to find this out after https://github.com/sig-net/mpc/pull/69 is merged.
- `Are we waiting on a lock held by a different thread?`
- `Or is the cryptographic algorithm itself not emitting more requests?`
